### ISSUE TITLE
App is crashing while calling platform specific code 

### ISSUE THREAD

**Issue by Faisalkc4u** [flutter/flutter#78225]
## Steps to Reproduce

1. Run `flutter create --template=plugin --platforms=macos macospermission`.
2. Create a plugin for macos swift
3. Write a platform code which returns a asynchronous
4. call the method

Sample Code: https://github.com/Faisalkc4u/macospermission.git
<details>
  <summary>Logs</summary>

```
flutter doctor -v
[✓] Flutter (Channel stable, 2.0.1, on macOS 11.2.3 20D91 darwin-x64, locale en-US)
    • Flutter version 2.0.1 at /usr/local/Caskroom/flutter/1.22.6/flutter
    • Framework revision c5a4b4029c (11 days ago), 2021-03-04 09:47:48 -0800
    • Engine revision 40441def69
    • Dart version 2.12.0

[✓] Android toolchain - develop for Android devices (Android SDK version 30.0.3)
    • Android SDK at /Users/mac/Library/Android/sdk
    • Platform android-30, build-tools 30.0.3
    • Java binary at: /Applications/Android Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b3-6915495)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 12.4, Build version 12D4e
    • CocoaPods version 1.10.1

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 4.1)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b3-6915495)

[✓] VS Code (version 1.54.1)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.20.0

[✓] Connected device (2 available)
    • macOS (desktop) • macos  • darwin-x64     • macOS 11.2.3 20D91 darwin-x64
    • Chrome (web)    • chrome • web-javascript • Google Chrome 89.0.4389.82

• No issues found!
```

</details>

**Comment by darshankawar** [comment]
@Faisalkc4u 
Cloned your repo but getting below error:

`Launching lib/main.dart on iPhone 12 Pro Max in debug mode...
No application found for TargetPlatform.ios.
Is your project missing an ios/Runner/Info.plist?
Consider running "flutter create ." to create one.
`

**Comment by Faisalkc4u** [comment]
This issue is for mac

**Comment by darshankawar** [comment]
@Faisalkc4u Thanks. that was my bad.

I ran the `example` directory on macOS and saw below:

<img width="959" alt="Screenshot 2021-03-16 at 6 42 54 PM" src="https://user-images.githubusercontent.com/67046386/111314716-ab71c900-8687-11eb-95b1-54327cbfbf39.png">

When do you exactly see the crash ? 

<details>
<summary>flutter doctor -v</summary>

```
[✓] Flutter (Channel stable, 2.0.2, on Mac OS X 10.15.4 19E2269 darwin-x64,
    locale en-GB)
    • Flutter version 2.0.2 at /Users/dhs/documents/Fluttersdk/flutter
    • Framework revision 8962f6dc68 (4 days ago), 2021-03-11 13:22:20 -0800
    • Engine revision 5d8bf811b3
    • Dart version 2.12.1

[!] Xcode - develop for iOS and macOS
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 12.3, Build version 12C33
    ! CocoaPods 1.9.3 out of date (1.10.0 is recommended).
        CocoaPods is used to retrieve the iOS and macOS platform side's plugin
        code that responds to your plugin usage on the Dart side.
        Without CocoaPods, plugins will not work on iOS or macOS.
        For more info, see https://flutter.dev/platform-plugins
      To upgrade see
      https://guides.cocoapods.org/using/getting-started.html#installation for
      instructions.

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] VS Code (version 1.54.1)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.18.1

[✓] Connected device (3 available)
    • iPhone 12 Pro Max (mobile) • A5473606-0213-4FD8-BA16-553433949729 • ios
      • com.apple.CoreSimulator.SimRuntime.iOS-14-3 (simulator)
    • macOS (desktop)            • macos                                •
      darwin-x64     • Mac OS X 10.15.4 19E2269 darwin-x64
    • Chrome (web)               • chrome                               •
      web-javascript • Google Chrome 89.0.4389.82

! Doctor found issues in 1 category.

```

</details>

**Comment by Faisalkc4u** [comment]
Click on "permission" button
which invokes a native method asking for permission 
before asking the permission the app is crashing


If you are opening example-> macOS in Xcode and running the project, After clicking on permission it's working fine.

**Comment by Faisalkc4u** [comment]
I have tested this code with following version and it's crashing 
macos 10.15.7 (Build 19H1016)
macos 11.2.3 (Build 20D91)

**Comment by darshankawar** [comment]
@Faisalkc4u 
Ran the example project on macOS using AS on latest stable and tapping on `permission` button isn't crashing for me.


https://user-images.githubusercontent.com/67046386/111428366-66987180-871d-11eb-89af-4ec69e62fbf6.mov


May be I am on a lesser OS version than you mentioned above ?

<details>
<summary>flutter doctor -v</summary>

```
[✓] Flutter (Channel stable, 2.0.2, on Mac OS X 10.15.4 19E2269 darwin-x64,
    locale en-GB)
    • Flutter version 2.0.2 at /Users/dhs/documents/Fluttersdk/flutter
    • Framework revision 8962f6dc68 (5 days ago), 2021-03-11 13:22:20 -0800
    • Engine revision 5d8bf811b3
    • Dart version 2.12.1

[!] Xcode - develop for iOS and macOS
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 12.3, Build version 12C33
    ! CocoaPods 1.9.3 out of date (1.10.0 is recommended).
        CocoaPods is used to retrieve the iOS and macOS platform side's plugin
        code that responds to your plugin usage on the Dart side.
        Without CocoaPods, plugins will not work on iOS or macOS.
        For more info, see https://flutter.dev/platform-plugins
      To upgrade see
      https://guides.cocoapods.org/using/getting-started.html#installation for
      instructions.

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] VS Code (version 1.54.1)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.18.1

[✓] Connected device (3 available)
    • iPhone 12 Pro Max (mobile) • A5473606-0213-4FD8-BA16-553433949729 • ios
      • com.apple.CoreSimulator.SimRuntime.iOS-14-3 (simulator)
    • macOS (desktop)            • macos                                •
      darwin-x64     • Mac OS X 10.15.4 19E2269 darwin-x64
    • Chrome (web)               • chrome                               •
      web-javascript • Google Chrome 89.0.4389.82

! Doctor found issues in 1 category.

```

</details>

**Comment by Faisalkc4u** [comment]
Your video says that you are using Mac OS version 10.15.4
You are using a lower than than mine
I'm getting problems on Mac OS 10.15.7 and 11.2.3

Here is the video for Mac OS 11.2.3


https://user-images.githubusercontent.com/12441355/111429834-f684ee00-8712-11eb-9fb6-bcfb1ad62d41.mov

**Comment by darshankawar** [comment]
Tagging @pedromassangocode  to verify this on 11.2.2.

**Comment by pedromassangocode** [comment]
I can reproduce this on latest version of MacOS (not sure if this only happens on latest version).

While running with VS Code the reproduces the issue: the app crash after pressing the button.
While running on Terminal app (non VS code or IntelliJ): the app don't crash, the permission dialog appear and you are able to accept it (see the logs bellow).


<details>
<summary>logs</summary>

```bash
pedromassango@MacBook-Pro-de-MacBook example % flutters run -d macos
Launching lib/main.dart on macOS in debug mode...
Building macOS application...                                           
Syncing files to device macOS...                                   161ms

Flutter run key commands.
r Hot reload. 🔥🔥🔥
R Hot restart.
h Repeat this help message.
d Detach (terminate "flutter run" but leave application running).
c Clear the screen
q Quit (terminate the application on the device).
An Observatory debugger and profiler on macOS is available at:
http://127.0.0.1:53101/zZM0w_B89Nk=/

Flutter DevTools, a Flutter debugger and profiler, on macOS is available at:
http://127.0.0.1:9104?uri=http%3A%2F%2F127.0.0.1%3A53101%2FzZM0w_B89Nk%3D%2F

Running with unsound null safety
For more information see https://dart.dev/null-safety/unsound-null-safety
flutter: notDetermined
flutter: authorized
flutter: authorized
```
</details>

Running on IntelliJ IDEA: the app simple accept the permission and no MacOS permission dialog is shown (see logs bellow)


<details>
<summary>logs</summary>

```bash
MacBook-Pro-de-MacBook:example pedromassango$ flutters run -d macos
Launching lib/main.dart on macOS in debug mode...
Building macOS application...                                           
Syncing files to device macOS...                                    78ms

Flutter run key commands.
r Hot reload. 🔥🔥🔥
R Hot restart.
h Repeat this help message.
d Detach (terminate "flutter run" but leave application running).
c Clear the screen
q Quit (terminate the application on the device).
An Observatory debugger and profiler on macOS is available at:
http://127.0.0.1:52995/mUOgcpwuKsQ=/

Flutter DevTools, a Flutter debugger and profiler, on macOS is available at:
http://127.0.0.1:9103?uri=http%3A%2F%2F127.0.0.1%3A52995%2FmUOgcpwuKsQ%3D%2F

Running with unsound null safety
For more information see https://dart.dev/null-safety/unsound-null-safety
flutter: authorized
flutter: authorized
flutter: authorized
flutter: authorized
flutter: authorized
flutter: authorized
flutter: authorized
```
</details>


This seems to be related to Mac permission behavior, as I've tried to revoke permission and running the app again (from terminal) it simply deny the permission request.

### Revoked permission for Terminal
![Screen Shot 2021-03-17 at 10 10 25](https://user-images.githubusercontent.com/67424416/111442780-00a1ef00-8709-11eb-89cd-60436e79cb37.png)

Now, running the app again the permission request is simply denied

<details>
<summary>logs</summary>

```bash
pedromassango@MacBook-Pro-de-MacBook example % flutters run -d macos
Launching lib/main.dart on macOS in debug mode...
Building macOS application...                                           
Syncing files to device macOS...                                    71ms

Flutter run key commands.
r Hot reload. 🔥🔥🔥
R Hot restart.
h Repeat this help message.
d Detach (terminate "flutter run" but leave application running).
c Clear the screen
q Quit (terminate the application on the device).
An Observatory debugger and profiler on macOS is available at:
http://127.0.0.1:53705/tWOM8D_R9QI=/

Flutter DevTools, a Flutter debugger and profiler, on macOS is available at:
http://127.0.0.1:9105?uri=http%3A%2F%2F127.0.0.1%3A53705%2FtWOM8D_R9QI%3D%2F

Running with unsound null safety
For more information see https://dart.dev/null-safety/unsound-null-safety
flutter: denied
flutter: denied
flutter: denied
flutter: denied
```
</details>


## VS Code

Since VS Code (or its Terminal) is not allowed to have Microphone permission, then it simply crash. 


<details>
<summary>logs</summary>

```bash
pedromassango@MacBook-Pro-de-MacBook example % flutters run -d macos
Launching lib/main.dart on macOS in debug mode...
Building macOS application...                                           
Syncing files to device macOS...                                    97ms

Flutter run key commands.
r Hot reload. 🔥🔥🔥
R Hot restart.
h Repeat this help message.
d Detach (terminate "flutter run" but leave application running).
c Clear the screen
q Quit (terminate the application on the device).
An Observatory debugger and profiler on macOS is available at: http://127.0.0.1:54209/ZHBpgokqD4g=/

Flutter DevTools, a Flutter debugger and profiler, on macOS is available at:
http://127.0.0.1:9107?uri=http%3A%2F%2F127.0.0.1%3A54209%2FZHBpgokqD4g%3D%2F

Running with unsound null safety
For more information see https://dart.dev/null-safety/unsound-null-safety
flutter: notDetermined
Lost connection to device.
pedromassango@MacBook-Pro-de-MacBook example % 
```
</details>

This seems to be more an VS Code issue so I would suggest you to file an issue in the related repository.

Following the indication of [VS Code Flutter Extension](https://marketplace.visualstudio.com/items?itemName=Dart-Code.flutter), the correct way to address the problem is to open an issue in the dedicated [repository](https://github.com/Dart-Code/Dart-Code/issues).
Closing, as this isn't an issue with Flutter itself. If you disagree, please write in the comments and I will reopen it.
Thank you

**Comment by dam-ease** [comment]
I am currently unable to reproduce this issue on the latest `stable` and `master` channels. I'd be closing this because it seems to have been fixed in the latest versions, if anybody disagrees, kindly write in comments and I'll reopen it or file a new issue using latest version code.


<details> <summary> flutter doctor -v </summary>

```bash
[✓] Flutter (Channel master, 3.11.0-2.0.pre.55, on macOS 13.3 22E252 darwin-arm64, locale en-NG)
    • Flutter version 3.11.0-2.0.pre.55 on channel master at /Users/dammya/fvm/versions/master
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision 835b892d7f (9 days ago), 2023-05-07 01:30:22 +0530
    • Engine revision 23f730efbf
    • Dart version 3.1.0 (build 3.1.0-83.0.dev)
    • DevTools version 2.23.1

[✓] Android toolchain - develop for Android devices (Android SDK version 32.1.0-rc1)
    • Android SDK at /Users/dammya/Library/Android/sdk
    • Platform android-33, build-tools 32.1.0-rc1
    • ANDROID_HOME = /Users/dammya/Library/Android/sdk
    • Java binary at: /Applications/Android Studio.app/Contents/jre/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 11.0.12+0-b1504.28-7817840)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 14.3)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Build 14E222b
    • CocoaPods version 1.11.3

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 2021.2)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 11.0.12+0-b1504.28-7817840)

[✓] IntelliJ IDEA Community Edition (version 2022.3.2)
    • IntelliJ at /Applications/IntelliJ IDEA CE.app
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart

[✓] VS Code (version 1.78.1)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.50.0

[✓] VS Code (version 1.78.0)
    • VS Code at /Users/dammya/.Trash/Visual Studio Code 20.07.05.app/Contents
    • Flutter extension version 3.50.0

[✓] Connected device (3 available)
    • Damilola’s iPhone (mobile) • 71c843fa5f3dec767999f36224653c0faa78660e • ios            • iOS 16.4 20E247
    • macOS (desktop)            • macos                                    • darwin-arm64   • macOS 13.3 22E252 darwin-arm64
    • Chrome (web)               • chrome                                   • web-javascript • Google Chrome 113.0.5672.92

[✓] Network resources
    • All expected network resources are available.

• No issues found!

```
```bash
[✓] Flutter (Channel stable, 3.10.0, on macOS 13.3 22E252 darwin-arm64, locale
    en-NG)
    • Flutter version 3.10.0 on channel stable at /Users/dammya/sdks/flutter
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision 84a1e904f4 (6 days ago), 2023-05-09 07:41:44 -0700
    • Engine revision d44b5a94c9
    • Dart version 3.0.0
    • DevTools version 2.23.1

[✓] Android toolchain - develop for Android devices (Android SDK version
    32.1.0-rc1)
    • Android SDK at /Users/dammya/Library/Android/sdk
    • Platform android-33, build-tools 32.1.0-rc1
    • ANDROID_HOME = /Users/dammya/Library/Android/sdk
    • Java binary at: /Applications/Android
      Studio.app/Contents/jre/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build
      11.0.12+0-b1504.28-7817840)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 14.3)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Build 14E222b
    • CocoaPods version 1.11.3

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 2021.2)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build
      11.0.12+0-b1504.28-7817840)

[✓] IntelliJ IDEA Community Edition (version 2022.3.2)
    • IntelliJ at /Applications/IntelliJ IDEA CE.app
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart

[✓] VS Code (version 1.78.1)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.50.0

[✓] VS Code (version 1.78.0)
    • VS Code at /Users/dammya/.Trash/Visual Studio Code 20.07.05.app/Contents
    • Flutter extension version 3.50.0

[✓] Connected device (3 available)
    • Damilola’s iPhone (mobile) • 71c843fa5f3dec767999f36224653c0faa78660e •
      ios            • iOS 16.4 20E247
    • macOS (desktop)            • macos                                    •
      darwin-arm64   • macOS 13.3 22E252 darwin-arm64
    • Chrome (web)               • chrome                                   •
      web-javascript • Google Chrome 113.0.5672.92

[✓] Network resources
    • All expected network resources are available.

• No issues found!
```
</details>
